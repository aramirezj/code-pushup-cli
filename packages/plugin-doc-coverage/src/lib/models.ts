export type UndocumentedItem = {
  file: string;
  type: string;
  name: string;
  line: number;
  class?: string;
};
