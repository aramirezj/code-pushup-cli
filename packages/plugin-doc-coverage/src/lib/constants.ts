import type { Audit, Group } from '@code-pushup/models';
import type { AuditSlug } from './models';

export const PLUGIN_SLUG = 'doc-coverage';

export const AUDITS_MAP: Record<AuditSlug, Audit> = {
  'classes-coverage': {
    slug: 'classes-coverage',
    title: 'Classes coverage',
    description: 'Coverage of classes',
  },
  'methods-coverage': {
    slug: 'methods-coverage',
    title: 'Methods coverage',
    description: 'Coverage of methods',
  },
  'functions-coverage': {
    slug: 'functions-coverage',
    title: 'Functions coverage',
    description: 'Coverage of functions',
  },
  'interfaces-coverage': {
    slug: 'interfaces-coverage',
    title: 'Interfaces coverage',
    description: 'Coverage of interfaces',
  },
  'variables-coverage': {
    slug: 'variables-coverage',
    title: 'Variables coverage',
    description: 'Coverage of variables',
  },
  'properties-coverage': {
    slug: 'properties-coverage',
    title: 'Properties coverage',
    description: 'Coverage of properties',
  },
  'types-coverage': {
    slug: 'types-coverage',
    title: 'Types coverage',
    description: 'Coverage of types',
  },
  'enums-coverage': {
    slug: 'enums-coverage',
    title: 'Enums coverage',
    description: 'Coverage of enums',
  },
} as const;

export const groups: Group[] = [
  {
    slug: 'documentation-coverage',
    title: 'Documentation coverage',
    description: 'Documentation coverage',
    refs: Object.keys(AUDITS_MAP).map(slug => {
      switch (slug as AuditSlug) {
        case 'classes-coverage':
        case 'functions-coverage':
        case 'methods-coverage':
          return { slug, weight: 2 };
        case 'interfaces-coverage':
        case 'properties-coverage':
        case 'types-coverage':
        default:
          return { slug, weight: 1 };
      }
    }),
  },
];
