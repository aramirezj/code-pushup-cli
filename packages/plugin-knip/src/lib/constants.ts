import { IssueType as KnipIssueType } from 'knip/dist/types/issues';

/**
 * @description
 * types that contain a knip `IssueSet`.
 */
export const ISSUE_SET_TYPES = ['files'] as const satisfies KnipIssueType[];

/**
 * @description
 * types that contain a knip `Issue`
 */
export const ISSUE_RECORDS_TYPES = [
  'dependencies',
  'devDependencies',
  'optionalPeerDependencies',
  'unlisted',
  'binaries',
  'unresolved',
  'exports',
  'nsExports',
  'types',
  'nsTypes',
  'enumMembers',
  'classMembers',
  'duplicates',
] as const satisfies KnipIssueType[];

export const ISSUE_TYPES = [
  ...ISSUE_SET_TYPES,
  ...ISSUE_RECORDS_TYPES,
] as const satisfies KnipIssueType[];
