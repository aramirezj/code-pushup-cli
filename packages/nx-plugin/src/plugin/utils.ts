import type { CreateNodesContext, ProjectConfiguration } from '@nx/devkit';
import { readFile } from 'node:fs/promises';
import * as path from 'node:path';
import { dirname, join } from 'node:path';
import { CP_TARGET_NAME } from './constants';
import { createTargets } from './target/targets';
import type { CreateNodesOptions, NormalizedCreateNodesOptions } from './types';

export async function normalizedCreateNodesContext(
  context: CreateNodesContext,
  projectConfigurationFile: string,
  createOptions: CreateNodesOptions = {},
) {
  const projectRoot = path.dirname(projectConfigurationFile);
}

export function normalizeCreateNodesOptions(
  options: unknown = {},
): NormalizedCreateNodesOptions {
  const { targetName = CP_TARGET_NAME } = options as CreateNodesOptions;
  return {
    ...(options as CreateNodesOptions),
    targetName,
  };
}

export async function loadProjectConfiguration(
  projectConfigurationFile: string,
): Promise<ProjectConfiguration> {
  const projectConfiguration = (await readFile(
    join(process.cwd(), projectConfigurationFile),
    'utf8',
  ).then(JSON.parse)) as Omit<ProjectConfiguration, 'root'> & { root?: string };
  if (
    !('name' in projectConfiguration) ||
    typeof projectConfiguration.name !== 'string'
  ) {
    throw new Error('Project name is required');
  }
  return {
    ...projectConfiguration,
    root: projectConfiguration.root ?? dirname(projectConfigurationFile),
  };
}

export async function createProjectConfiguration(
  projectConfiguration: ProjectConfiguration,
  options: CreateNodesOptions,
): Promise<
  Pick<ProjectConfiguration, 'targets'> &
    Partial<Pick<ProjectConfiguration, 'namedInputs'>>
> {
  const normalizeOptions = normalizeCreateNodesOptions(options);
  return {
    namedInputs: {},
    targets: await createTargets(projectConfiguration, normalizeOptions),
  };
}
