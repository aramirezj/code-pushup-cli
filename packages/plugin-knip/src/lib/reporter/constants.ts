import { ISSUE_TYPES } from '../constants';

export const KNIP_REPORT_NAME = 'knip-report.json';
export const KNIP_RAW_REPORT_NAME = 'knip-raw-report.json';

export const ISSUE_TYPE_TITLE: Record<
  (typeof ISSUE_TYPES)[number] | '_files',
  string
> = {
  files: 'Unused files',
  _files: 'Unused files',
  dependencies: 'Unused dependencies',
  devDependencies: 'Unused devDependencies',
  optionalPeerDependencies: 'Referenced optional peerDependencies',
  unlisted: 'Unlisted dependencies',
  binaries: 'Unlisted binaries',
  unresolved: 'Unresolved imports',
  exports: 'Unused exports',
  nsExports: 'Exports in used namespace',
  types: 'Unused exported types',
  nsTypes: 'Exported types in used namespace',
  enumMembers: 'Unused exported enum members',
  classMembers: 'Unused exported class members',
  duplicates: 'Duplicate exports',
} as const;

export const ISSUE_TYPE_MESSAGE: Record<
  (typeof ISSUE_TYPES)[number] | '_files',
  (arg: string) => string
> = {
  files: (file: string) => `Unused file ${file}`,
  // eslint-disable-next-line  @typescript-eslint/naming-convention
  _files: (file: string) => `Unused file ${file}`,
  dependencies: (dep: string) => `Unused dependency ${dep}`,
  devDependencies: (dep: string) => `Unused devDependency ${dep}`,
  optionalPeerDependencies: (dep: string) =>
    `Referenced optional peerDependency ${dep}`,
  unlisted: (dep: string) => `Unlisted dependency ${dep}`,
  binaries: (binary: string) => `Unlisted binary ${binary}`,
  unresolved: (importName: string) => `Unresolved import ${importName}`,
  exports: (exportName: string) => `Unused export ${exportName}`,
  nsExports: (namespace: string) => `Exports in used namespace ${namespace}`,
  types: (type: string) => `Unused exported type ${type}`,
  nsTypes: (namespace: string) =>
    `Exported types in used namespace ${namespace}`,
  enumMembers: (enumMember: string) =>
    `Unused exported enum member ${enumMember}`,
  classMembers: (classMember: string) =>
    `Unused exported class member ${classMember}`,
  duplicates: (duplicate: string) => `Duplicate export ${duplicate}`,
} as const;
