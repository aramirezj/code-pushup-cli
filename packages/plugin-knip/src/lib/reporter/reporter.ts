import type { ReporterOptions } from 'knip';
import { writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';
import { ensureDirectoryExists, ui } from '@code-pushup/utils';
import { KNIP_REPORT_NAME } from './constants';
import { parseCustomReporterOptions } from './model';
import { knipToCpReport } from './utils';

/**
 * @description
 * This custom knip reporter produces code-pushup AuditOutputs and saves it to the filesystem
 *
 * @example
 * run the following command to test it:
 * npx knip --reporter ./dist/packages/plugin-knip/reporter.js --reporter-options='{\"outputFile\":\"my-knip-report.json\"}'
 *
 */
export const knipReporter = async (knipReporterOptions: ReporterOptions) => {
  const { options, report, issues } = knipReporterOptions;
  const customReporterOptions = parseCustomReporterOptions(options);
  const {
    verbose,
    outputFile = KNIP_REPORT_NAME,
    rawOutputFile,
  } = customReporterOptions;
  if (verbose) {
    ui().logger.info(
      `Reporter called with options: ${JSON.stringify(
        customReporterOptions,
        null,
        2,
      )}`,
    );
  }
  if (rawOutputFile != null) {
    await ensureDirectoryExists(dirname(rawOutputFile));
    await writeFile(
      rawOutputFile,
      JSON.stringify(
        {
          ...knipReporterOptions,
          issues: {
            ...issues,
            files: [...issues.files], // files is a Set<string>
          },
          options: customReporterOptions,
        },
        null,
        2,
      ),
    );
    if (verbose) {
      ui().logger.info(`Saved raw report to ${rawOutputFile}`);
    }
  }

  const result = await knipToCpReport({ issues, report });

  await ensureDirectoryExists(dirname(outputFile));
  await writeFile(outputFile, JSON.stringify(result, null, 2));
  if (verbose) {
    ui().logger.info(`Saved report to ${outputFile}`);
  }
};
