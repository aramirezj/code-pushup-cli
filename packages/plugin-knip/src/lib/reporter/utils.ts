import type {
  IssueType,
  Issue as KnipIssue,
  IssueRecords as KnipIssueRecords,
  IssueSet as KnipIssueSet,
  Issues as KnipIssues,
  IssueSeverity as KnipSeverity,
  ReporterOptions,
} from 'knip/dist/types/issues';
import {
  AuditOutput,
  AuditOutputs,
  IssueSeverity as CondPushupIssueSeverity,
  Issue as CpIssue,
} from '@code-pushup/models';
import { formatGitPath, getGitRoot, slugify } from '@code-pushup/utils';
import { ISSUE_RECORDS_TYPES, ISSUE_SET_TYPES } from '../constants';
import { ISSUE_TYPE_MESSAGE, ISSUE_TYPE_TITLE } from './constants';

const severityMap: Record<KnipSeverity | 'unknown', CondPushupIssueSeverity> = {
  unknown: 'info',
  off: 'info',
  error: 'error',
  warn: 'warning',
} as const;

export function knipIssueSetToIssues(
  issueType: (typeof ISSUE_SET_TYPES)[number],
  knipIssueSetAsArray: string[],
  gitRoot: string,
): CpIssue[] {
  return knipIssueSetAsArray.map(
    (filePath): CpIssue => ({
      message: ISSUE_TYPE_MESSAGE[issueType](formatGitPath(filePath, gitRoot)),
      severity: severityMap['unknown'], // @TODO rethink
      source: {
        file: formatGitPath(filePath, gitRoot),
      },
    }),
  );
}

export function getPosition(
  issue: Pick<KnipIssue, 'symbols' | 'line' | 'col'>,
):
  | false
  | {
      startColumn: number;
      startLine: number;
    } {
  const hasSymbols = issue.symbols?.length ?? 0;
  if (hasSymbols) {
    // we want to have the line of code of the last symbol present in a file
    const { col = -1, line = -1 } = issue.symbols?.at(-1) ?? {};
    return { startColumn: col, startLine: line };
  }
  if (issue.line && issue.col) {
    return {
      startColumn: issue.col,
      startLine: issue.line,
    };
  }
  return false;
}

export function knipIssueToIssue(
  knipIssues: (Pick<
    KnipIssue,
    'symbol' | 'filePath' | 'severity' | 'line' | 'col'
  > & { type: (typeof ISSUE_RECORDS_TYPES)[number] })[],
  gitRoot: string,
): CpIssue[] {
  return knipIssues.map((issue): CpIssue => {
    const { type, symbol, filePath, severity = 'unknown' } = issue;
    const position = getPosition(issue);
    return {
      message: ISSUE_TYPE_MESSAGE[type](symbol),
      severity: severityMap[severity],
      source: {
        file: formatGitPath(filePath, gitRoot),
        ...(position ? { position } : {}),
      },
    };
  });
}

export async function toIssues(
  issueType: IssueType,
  issues: KnipIssues,
): Promise<CpIssue[]> {
  const isSet = issues[issueType] instanceof Set;
  const issuesForType: string[] | KnipIssue[] = isSet
    ? [...(issues[issueType] as KnipIssueSet)]
    : Object.values(issues[issueType] as KnipIssueRecords).flatMap(
        Object.values,
      );

  const gitRoot = await getGitRoot();
  if (issuesForType.length > 0) {
    if (isSet) {
      const knipIssueSetPaths = issuesForType as string[];
      return knipIssueSetToIssues(
        issueType as (typeof ISSUE_SET_TYPES)[number],
        knipIssueSetPaths,
        gitRoot,
      );
    } else {
      const knipIssues = issuesForType as KnipIssue[];
      return knipIssueToIssue(knipIssues, gitRoot);
    }
  }
  return [];
}

export function knipToCpReport({
  issues: rawIssues,
  report,
}: Pick<ReporterOptions, 'report' | 'issues'>): Promise<AuditOutputs> {
  return Promise.all(
    Object.entries(report)
      .filter(([_, isReportType]) => isReportType)
      .map(async ([issueType]): Promise<AuditOutput> => {
        const issues = await toIssues(issueType as IssueType, rawIssues);

        return {
          slug: slugify(ISSUE_TYPE_TITLE[issueType as IssueType]),
          score: issues.length === 0 ? 1 : 0,
          value: issues.length,
          ...(issues.length > 0 ? { details: { issues } } : {}),
        };
      }),
  );
}
