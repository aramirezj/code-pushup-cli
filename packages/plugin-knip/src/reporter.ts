import { knipReporter } from './lib/reporter';

export default knipReporter;
