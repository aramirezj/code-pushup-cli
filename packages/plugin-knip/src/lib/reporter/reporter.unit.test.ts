import type { ReporterOptions } from 'knip';
import { IssueRecords, IssueSet } from 'knip/dist/types/issues';
import { fs as memfsFs } from 'memfs';
import { join } from 'node:path';
import { describe, expect, it, vi } from 'vitest';
import { AuditOutputs } from '@code-pushup/models';
import { MEMFS_VOLUME, getLogMessages } from '@code-pushup/test-utils';
import { ui } from '@code-pushup/utils';
import { rawReport } from '../../../mocks/fixtures/raw-knip.report';
import { KNIP_RAW_REPORT_NAME, KNIP_REPORT_NAME } from './constants';
import { CustomReporterOptions } from './model';
import { knipReporter } from './reporter';

vi.mock('@code-pushup/utils', async () => {
  const actual = await vi.importActual('@code-pushup/utils');
  return {
    ...actual,
    getGitRoot: vi
      .fn()
      .mockResolvedValue('/Users/username/Projects/code-pushup-cli/'),
  };
});

describe('knipReporter', () => {
  it('should saves report to file system by default', async () => {
    await expect(
      knipReporter({
        report: {
          files: true,
        },
        issues: {
          files: new Set(['main.js']),
        },
      } as ReporterOptions),
    ).resolves.toBeUndefined();

    expect(getLogMessages(ui().logger)).toHaveLength(0);
  });

  it('should accept reporter option outputFile', async () => {
    const outputFile = 'my-report.json';
    await expect(
      knipReporter({
        report: {
          files: true,
        },
        issues: {
          files: new Set(['main.js']),
        },
        options: JSON.stringify({ outputFile } satisfies CustomReporterOptions),
      } as ReporterOptions),
    ).resolves.toBeUndefined();

    expect(getLogMessages(ui().logger)).toHaveLength(0);

    const auditOutputs = JSON.parse(
      (
        await memfsFs.promises.readFile(outputFile, { encoding: 'utf8' })
      ).toString(),
    );
    expect(auditOutputs).toStrictEqual([
      expect.objectContaining({ slug: 'unused-files' }),
    ]);
  });

  it('should accept reporter option rawOutputFile', async () => {
    const rawOutputFile = KNIP_RAW_REPORT_NAME;
    const fileIssueSet: IssueSet = new Set(['main.js']);
    const unlistedIssueRecords: IssueRecords = {
      '/User/username/code-pushup-cli/packages/utils/.eslintrc.json': {
        'jsonc-eslint-parser': {
          type: 'unlisted',
          symbol: 'jsonc-eslint-parser',
          filePath:
            '/User/username/code-pushup-cli/packages/utils/package.json',
        },
      },
    };

    await expect(
      knipReporter({
        report: {
          files: true,
          unlisted: true,
        },
        issues: {
          files: fileIssueSet,
          unlisted: unlistedIssueRecords,
        },
        options: JSON.stringify({ rawOutputFile }),
        // other reporter options for debugging purpose
        counters: { files: 1, unlisted: 1 },
      } as ReporterOptions),
    ).resolves.toBeUndefined();

    expect(getLogMessages(ui().logger)).toHaveLength(0);

    const rawKnipReport = JSON.parse(
      (
        await memfsFs.promises.readFile(rawOutputFile, { encoding: 'utf8' })
      ).toString(),
    );
    expect(rawKnipReport.report).toStrictEqual({ files: true, unlisted: true });
    expect(rawKnipReport.options).toStrictEqual({ rawOutputFile });
    expect(rawKnipReport.issues.files).toStrictEqual(['main.js']);
    expect(rawKnipReport.issues.unlisted).toStrictEqual(unlistedIssueRecords);
    expect(rawKnipReport.counters).toStrictEqual({ files: 1, unlisted: 1 });
  });

  it('should log if custom reporter option verbose is true', async () => {
    const reporterOptions: CustomReporterOptions = {
      verbose: true,
      outputFile: KNIP_REPORT_NAME,
      rawOutputFile: KNIP_RAW_REPORT_NAME,
    };
    await expect(
      knipReporter({
        report: { files: true },
        issues: { files: new Set(['main.js']) },
        options: JSON.stringify(reporterOptions),
      } as ReporterOptions),
    ).resolves.toBeUndefined();

    expect(getLogMessages(ui().logger)).toHaveLength(3);
    expect(getLogMessages(ui().logger).at(0)).toBe(
      `[ blue(info) ] Reporter called with options: ${JSON.stringify(
        reporterOptions,
        null,
        2,
      )}`,
    );
    expect(getLogMessages(ui().logger).at(1)).toBe(
      `[ blue(info) ] Saved raw report to ${reporterOptions.rawOutputFile}`,
    );
    expect(getLogMessages(ui().logger).at(2)).toBe(
      `[ blue(info) ] Saved report to ${reporterOptions.outputFile}`,
    );
  });

  it('should produce valid audit outputs', async () => {
    await expect(
      knipReporter(rawReport as ReporterOptions),
    ).resolves.toBeUndefined();

    const auditOutputsContent = await memfsFs.promises.readFile(
      join(MEMFS_VOLUME, KNIP_REPORT_NAME),
      { encoding: 'utf8' },
    );
    const auditOutputsJson = JSON.parse(
      auditOutputsContent.toString(),
    ) as AuditOutputs;
    expect(auditOutputsJson).toMatchSnapshot();
  });
});
