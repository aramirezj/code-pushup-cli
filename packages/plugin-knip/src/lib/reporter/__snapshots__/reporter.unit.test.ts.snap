// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`knipReporter > should produce valid audit outputs 1`] = `
[
  {
    "details": {
      "issues": [
        {
          "message": "Unused file code-pushup.config.ts",
          "severity": "info",
          "source": {
            "file": "code-pushup.config.ts",
          },
        },
      ],
    },
    "score": 0,
    "slug": "unused-files",
    "value": 1,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Unused dependency cli-table3",
          "severity": "error",
          "source": {
            "file": "package.json",
          },
        },
      ],
    },
    "score": 0,
    "slug": "unused-dependencies",
    "value": 1,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Unused devDependency @trivago/prettier-plugin-sort-imports",
          "severity": "error",
          "source": {
            "file": "package.json",
          },
        },
      ],
    },
    "score": 0,
    "slug": "unused-devdependencies",
    "value": 1,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Unused devDependency ts-node",
          "severity": "error",
          "source": {
            "file": "package.json",
          },
        },
      ],
    },
    "score": 0,
    "slug": "referenced-optional-peerdependencies",
    "value": 1,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Unlisted dependency jsonc-eslint-parser",
          "severity": "info",
          "source": {
            "file": "../../../../User/username/code-pushup-cli/packages/utils/package.json",
          },
        },
        {
          "message": "Unlisted dependency jsonc-eslint-parser",
          "severity": "info",
          "source": {
            "file": "../../../../User/username/code-pushup-cli/packages/utils/package.json",
          },
        },
      ],
    },
    "score": 0,
    "slug": "unlisted-dependencies",
    "value": 2,
  },
  {
    "score": 1,
    "slug": "unlisted-binaries",
    "value": 0,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Unresolved import smo-package",
          "severity": "info",
          "source": {
            "file": "packages/models/src/lib/category-config.ts",
            "position": {
              "startColumn": 23,
              "startLine": 8,
            },
          },
        },
      ],
    },
    "score": 0,
    "slug": "unresolved-imports",
    "value": 1,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Unused export duplicateErrorMsg",
          "severity": "error",
          "source": {
            "file": "packages/models/src/lib/category-config.ts",
            "position": {
              "startColumn": 17,
              "startLine": 54,
            },
          },
        },
      ],
    },
    "score": 0,
    "slug": "unused-exports",
    "value": 1,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Unused exported type GroupMeta",
          "severity": "error",
          "source": {
            "file": "packages/models/src/lib/group.ts",
            "position": {
              "startColumn": 13,
              "startLine": 26,
            },
          },
        },
      ],
    },
    "score": 0,
    "slug": "unused-exported-types",
    "value": 1,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Unused exported enum member unusedMember",
          "severity": "error",
          "source": {
            "file": "packages/models/src/lib/group.ts",
            "position": {
              "startColumn": 13,
              "startLine": 26,
            },
          },
        },
      ],
    },
    "score": 0,
    "slug": "unused-exported-enum-members",
    "value": 1,
  },
  {
    "details": {
      "issues": [
        {
          "message": "Duplicate export configurationGenerator|default",
          "severity": "error",
          "source": {
            "file": "../../../michael_hladky/WebstormProjects/quality-metrics-cli/packages/nx-plugin/src/generators/configuration/generator.ts",
            "position": {
              "startColumn": 15,
              "startLine": 54,
            },
          },
        },
      ],
    },
    "score": 0,
    "slug": "duplicate-exports",
    "value": 1,
  },
]
`;
